crates/sovereign-sdk/module-system/sov-state/src/codec/borsh_codec.rs

```diff
@@ -1,37 +1,98 @@
 
 /// A [`StateCodec`] that uses [`borsh`] for all keys and values.
-#[derive(Debug, Default, PartialEq, Eq, Clone, BorshDeserialize, borsh::BorshSerialize)]
+#[derive(Debug, Default, PartialEq, Eq, Clone)]
 pub struct BorshCodec;
 
-impl<K> StateKeyCodec<K> for BorshCodec
+impl StateKeyCodec<AlloyU256> for BorshCodec {
+    fn encode_key(&self, value: &AlloyU256) -> Vec<u8> {
+        borsh::to_vec(value.as_limbs()).expect("Failed to serialize key")
+    }
+}
+
+impl StateKeyCodec<AlloyAddress> for BorshCodec {
+    fn encode_key(&self, value: &AlloyAddress) -> Vec<u8> {
+        borsh::to_vec(value.as_slice()).expect("Failed to serialize key")
+    }
+}
+
+impl StateKeyCodec<ModuleAddress> for BorshCodec {
+    fn encode_key(&self, value: &ModuleAddress) -> Vec<u8> {
+        borsh::to_vec(&value).expect("Failed to serialize key")
+    }
+}
+
+impl StateValueCodec<AlloyU256> for BorshCodec {
+    type Error = std::io::Error;
+
+    fn encode_value(&self, value: &AlloyU256) -> Vec<u8> {
+        let t = value.as_limbs();
+        borsh::to_vec(t).unwrap()
+    }
+
+    fn try_decode_value(&self, bytes: &[u8]) -> Result<AlloyU256, Self::Error> {
+        let s: [u64; 4] = borsh::from_slice(bytes)?;
+        Ok(AlloyU256::from_limbs(s))
+    }
+}
+
+impl<T> StateKeyCodec<Vec<T>> for BorshCodec
 where
-    K: BorshSerialize + BorshDeserialize,
+    T: BorshSerialize,
 {
-    fn encode_key(&self, value: &K) -> Vec<u8> {
+    fn encode_key(&self, value: &Vec<T>) -> Vec<u8> {
         borsh::to_vec(value).expect("Failed to serialize key")
     }
 }
 
-impl<V> StateValueCodec<V> for BorshCodec
+impl<T> StateValueCodec<Vec<T>> for BorshCodec
 where
-    V: BorshSerialize + BorshDeserialize,
+    T: BorshSerialize + BorshDeserialize,
 {
     type Error = std::io::Error;
 
-    fn encode_value(&self, value: &V) -> Vec<u8> {
+    fn encode_value(&self, value: &Vec<T>) -> Vec<u8> {
         borsh::to_vec(value).expect("Failed to serialize value")
     }
 
-    fn try_decode_value(&self, bytes: &[u8]) -> Result<V, Self::Error> {
+    fn try_decode_value(&self, bytes: &[u8]) -> Result<Vec<T>, Self::Error> {
         borsh::from_slice(bytes)
     }
 }
 
+macro_rules! impl_borsh_codec {
+    ($t:tt) => {
+        impl StateKeyCodec<$t> for BorshCodec {
+            fn encode_key(&self, value: &$t) -> Vec<u8> {
+                borsh::to_vec(value).expect("Failed to serialize key")
+            }
+        }
+
+        impl StateValueCodec<$t> for BorshCodec {
+            type Error = std::io::Error;
+
+            fn encode_value(&self, value: &$t) -> Vec<u8> {
+                borsh::to_vec(value).expect("Failed to serialize value")
+            }
+
+            fn try_decode_value(&self, bytes: &[u8]) -> Result<$t, Self::Error> {
+                borsh::from_slice(bytes)
+            }
+        }
+    };
+}
+
+impl_borsh_codec!(usize);
+impl_borsh_codec!(i32);
+impl_borsh_codec!(u32);
+impl_borsh_codec!(u64);
+impl_borsh_codec!(String);
```

Comment:
can prealloc exact

Reply:
It already does

Reply:
Okay, I will check for String

Reply:
https://github.com/chainwayxyz/citrea/pull/1934